src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -492,6 +491,172 @@ public Task OpenAsync(CancellationToken token)
+        {
+            get { return this.innerDuplexChannel.RemoteAddress; }
+        }
+
+        public Uri Via
+        {
+            get { return this.innerDuplexChannel.Via; }
+        }
+
+        protected IDuplexChannel InnerDuplexChannel
+        {
+            get { return this.innerDuplexChannel; }
+        }
+
+        internal SecurityProtocol SecurityProtocol { get; set; }
+
+      //  public IReplyChannel OuterChannel { get; private set; }
+
+        public T GetProperty<T>() where T : class
+        {
+            T tObj = _serviceProvider.GetService<T>();
+            return tObj ?? InnerDuplexChannel.GetProperty<T>();
+        }
+
+        public abstract Task DispatchAsync(RequestContext context);
+        public abstract Task DispatchAsync(Message message);
+
+        public Task SendAsync(Message message, TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            message = SecurityProtocol.SecureOutgoingMessage (message, timeoutHelper.GetCancellationToken());
+            return InnerDuplexChannel.SendAsync(message, timeoutHelper.GetCancellationToken());
+        }
+    }
+
+    sealed class SecurityDuplexSessionChannelDispatcher : SecurityDuplexChannel<IDuplexSessionChannel>, IDuplexSessionChannel
+    {
+        bool sendUnsecuredFaults;
+
+        public SecurityDuplexSessionChannelDispatcher(SecurityServiceDispatcher serviceDispatcher, IDuplexSessionChannel innerChannel, SecurityProtocol securityProtocol, SecurityListenerSettingsLifetimeManager settingsLifetimeManager)
+            : base(serviceDispatcher, innerChannel, securityProtocol, settingsLifetimeManager)
+        {
+            // sendUnsecuredFaults = channelManager.SendUnsecuredFaults;
+            SecurityServiceDispatcher = serviceDispatcher;
+        }
+
+        public IDuplexSession Session
+        {
+            get { return ((IDuplexSessionChannel)InnerDuplexChannel).Session; }
+        }
+
+        public EndpointAddress LocalAddress => throw new NotImplementedException();
+
+        public IServiceChannelDispatcher ChannelDispatcher { get; set; }
+
+        public SecurityServiceDispatcher SecurityServiceDispatcher { get; }
+
+        public CommunicationState State => InnerDuplexChannel.State;
+
+        public event EventHandler Closed;
+        public event EventHandler Closing;
+        public event EventHandler Faulted;
+        public event EventHandler Opened;
+        public event EventHandler Opening;
+
+        public void Abort()
+        {
+            return;// Task.CompletedTask;
+        }
+
+        public Task CloseAsync()
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task CloseAsync(CancellationToken token)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task OpenAsync()
+        {
+            return Task.CompletedTask;
+           // return InnerDuplexChannel.OpenAsync();
```

Comment:
There's inconsistency here. This method is returning a completed task, the one below with a CancellationToken is throwing. And there is an inner channel which you could call open on. Also the CloseAsync call is also throwing. This needs tidying up and making consistent. Should these objects be having OpenAsync called on them?